Member SignUp/SignIn

yumkoori opened the pull request:
회원 가입 기능을 RestController로 구현함

yumkoori committed:
애플리케이션에 JoinCommand 추가

-원래 requestDto로 Repository에서 의존하게 되도록 되어있었음

yumkoori committed:
회원 가입전에 DB에서 회원 조회시 없을때 예외처리

-조회할때 회원을 찾지 못하면 일단 null을 반환하도록 했지만  예외를 던져서 Controller쪽에서 처리하도록 나중에 고쳐야함

yumkoori committed:
login 기능 완성

-test도 통과함.

yumkoori committed:
부트스트랩 라이브러리 추가

-로그인 기능을 웹에서 테스트해보기 위해서 부트스트랩 라이브러리를 추가했음.

yumkoori committed:
타임리프 관련 설정을 yml파일로 옮김

-보기 편하려고 yml 파일로 옮김
-나중에는 properties에 있는 모든 설정 정보를 yml로 바꿀예정

yumkoori committed:
mybatis관련 테스트용  파일들을 삭제

-test용으로 만들었던 mybatis관련 파일들 삭제
-test용 html을 삭제

yumkoori committed:
회원가입 기능을 웹에서 테스트하기 위해 html파일 등록

-회원가입 입력 폼 관련 html
-기본 welcomPage html

yumkoori committed:
git pr을 test하기 위해 일단 commit해보는 미완성 WebAdapter

-아직 미완성

yumkoori committed:
Feat : response를 위한 dto와 http 응답 메세지 클래스 작성

- ResultDto: 응답 httpp 메세지에 반환해줄 정보들을 작성함
- ResponseMemberDto : 회원가입 성공시 닉네임을 반환하도록 작성함

yumkoori committed:
Form Controller와 RestController를 따로 분리함

- 한쪽으로 몰아넣으니까 @Controller 따로, @RestController 따로 작성해줘야해서 분리하는게 나을것 같아서 분리함
- RestController는 test 완료

james9dev commented on src/main/java/toyProject/toyProject01/member/domain/Member.java:
```diff
@@ -9,7 +9,7 @@
 
     //비즈니스 엔티티
-    private final Long memberNumber;
+    private final Long Number;
```
회원 번호를 지칭 할때는 통상적으로 number 보다는 no나 memberNo를 사용하거나, id를 사용합니다.
아래 로그인용 id를 memberId로 지정해놓았는데, memberId는 통상적으로 회원가입시 생성된 유니크값을 지칭하는 경우가 많기 때문에 혼돈이 올수 있으니, loginId 또는 최근에는 대부분의 id는 email로 받으니, email로 해두면 좋을것 같습니다.

그리고, 일반적으로 변수명의 시작은 소문자로 하는게 좋습니다.
변수가 상수일 경우에는 대문자 스네이크 케이스로 하고, 일반 변수일 경우에는 소문자로 시작하는 카멜케이스로 합니다.

yumkoori replied:
수정했습니다. 실수로 몇개의 커밋 메세지에는 제목에 Type적는것을 누락했습니다. 다음부터는 빼먹지 않도록 하겠습니다.

james9dev commented on src/main/java/toyProject/toyProject01/member/adapter/out/persistence/MemberJpaAdapter.java:
```diff
@@ -36,13 +40,14 @@ public Member loadMemberWithNumber(Long memberNumber) {
+            return memberMapper.mapToDomainMember(findMember);
+        } catch (Exception e) {
+            log.info("해당 회원을 찾을 수 없습니다.");
+            return null;            //예외로 바꿔야함.
```
Optional로 바꾸고, 예외는 Service에서 처리해서 Controller로 전달하는게 좋을듯 합니다. 회원을 찾을수 없다는걸 결국은 Controller를 통해서 프론트까지 전달해야 되는데, 여기서 예외를 처리 하게 되면, Service에서 받고, 다시 또 Service용 예외를 만들어줘야 할텐데, 안그러면 Out Adapter에 선언된 예외를 In Adapter(Controller)가 알아야 하는 Out/In Adapter 계층간 의존성이 발생하는 문제가 생길거 같네요.

yumkoori replied:
SpringDataRepository에서 반환 타입을 Optional로 바꾸고, Adapter에서는 예외를 처리하지않고 Repository에서 넘어온 Optional 값이 비었는지만 판단해서 예외를 던지면 Service에서 예외를 처리해서 Service 전용 예외를 Controller로 전달해줄 수 있도록 변경했습니다.

yumkoori committed:
Refactor : 도메인 영역에서 memberId 대신에 email로 대체

- memberId필드는 email필드로 대체될 예정이기 때문에 삭제함
- member 생성자에 memberId필드를 삭제하고, 그자리에 email 필드의 순서를 첫번째로 변경하여 대체됨을 강조함.
- mapToMember 메서드에 변경된 member생성자에 정의된 순서와 규칙에 맞게 수정함.

yumkoori committed:
애플리케이션 영역에서 memberId를 email로 대체함

- LoadMemberPort 인터페이스에서
  number필드와 id필드를 더이상 사용하지 않고 email필드로 대체할 것이기 때문에
  loadMemberWithNumber메서드를 삭제하고, memberId로 조회하는 대신에 email로 조회하도록 수정함

- JoinCommand와 LoginCommand에서 email로 대체된 memberId필드를 삭제하고, email필드를 첫번째 순서로 수정함

- MemberService에서 수정된 loadMemberPort인터페이스를 적용하도록 메서드 이름을 수정하고, 인수도 memberId대신에 email을 넘겨주도록 수정함.

yumkoori committed:
영속성 영역에서 memberId를 email로 대체함

- 엔티티에서 number필드와 memberId필드를 emai필드로 대체하기 위해 삭제하고, PK값을 email필드로 지정하도록 수정함.

- 리포지토리에서 JpaRepository상속 타입에서 PK의 타입을 email타입인 String으로 변경하고, 쿼리에 id관련된 부분을 email로 수정함.

yumkoori committed:
Web어댑터 영역에서 memberId를 email로 대체함

- RequestJoinDto에서 memberId필드를 삭제하고 email필드로 자리를 교체함

- 변경된 joinCommand 생성자에 따라서 수정함.

yumkoori committed:
Test : 테스트 영역에서 memberId를 email로 대체함

yumkoori committed:
Refactor : Repository에서 회원 정보를 조회할때 반환 타입을 Optional로 수정함

- db 조회시 회원정보 조회 결과가 있을 수도 있고, 없을 수도 있기 때문에 반환타입을 optional로 수정함

yumkoori committed:
Refactor : 아웃어댑터에서 예외처리를 하지않고 서비스에서 예외를 처리하도록 변경함

- repository에서 회원 정보 조회시 optional타입이 반환됨
  만약 조회되는 회원정보가 없으면 서비스로 예외를 던지고,
  조회되는 회원 정보가 있으면 반환 타입이 해당객체 타입으로 반환되어 정상적인 로직이 수행됨

yumkoori committed:
Feat : 서비스 영역에서 사용될 전용 예외 클래스를 생성함

yumkoori committed:
Refactor : 회원가입 서비스에서 아웃어댑터의 예외를 처리하도록 수정함

- 회원데이터 조회시에 데이터가 없을 경우 넘어오는 예외를 서비스에서 처리하도록 try문을 추가함
- 회원정보가 이미 있을 경우에 컨트롤러에 회원가입 불가능 상태를 전달하기 위해서 서비스 전용 예외를 전달해 알려줌
- 회원정보가 조회되지 않아서 예외가 넘어오면 예외를 잡고, 회원가입 정상 로직을 실행시킴

yumkoori committed:
Refactor : 로그인 서비스에서 아웃어댑터의 예외를 처리하도록 수정함

- 회원데이터 조회시에 데이터가 없을 경우 넘어오는 예외를 서비스에서 처리하도록 try문을 추가함
- 회원정보가 예외없이 조회되면, 비밀번호가 일치하는지 확인해서 틀리면 서비스예외를 전달해줌
- 만약 어댑터에서 예외가 넘어오면 어댑터의 예외를 서비스예외로 변경해서 컨트롤러에 전달해줌
- 모든 검증이 끝나면 true를 반환해줌

yumkoori committed:
Test : 회원서비스의 테스트에서 실패케이스를 예외로 검증하도록 수정함

- 서비스 계층에서 로직 수행이 실패했을 경우 예외를 던져주도록 변경했기 때문에 테스트도 수정함
- 서비스 계층에서 예외가 넘어오면 서비스 전용 예외가 전달되는지 검증함

james9dev commented on src/main/java/toyProject/toyProject01/member/application/service/MemberService.java:
회원가입 함수 join 코드의 try catch 구분에서 실패 동작이 catch 구분에서 처리 되지 않고, try 구분에서 처리 되고 있는데, 코드 가독성이 떨어져서, 코드를 읽는데 어려움을 줄수 있으니, 회원가입 가능 여부만 체크하는 함수를 따로 만들어서 사용하면 코드 간결해질거 같네요.

```
//회원 가입 가능 여부를 체크하는 함수를 service 클래스에 만들어 두고 사용하면 코드 읽기가 편해질듯 합니다.
private bool possibleJoinWithEmail(String email) {
    try {
            Member findMember = loadMemberPort.loadMemberWithEmail(joinCommand.getEmail());
            if (findMember != null) {
                return false;
            }
    } catch (NoSuchElementException e) {
            log.info("중복되는 회원이 없습니다. 회원가입이 가능합니다.", e);
            return true;
    }
}

@Override
public boolean Join(JoinCommand joinCommand) {   
    
    if (possibleJoinWithEmail(joinCommand.getEmail())) {

        Member member = Member.mapToMember(joinCommand);
        saveMemberPort.saveMember(member);
        
        return true;
    }
    
    return false;
}
```

yumkoori replied:
수정완료 했습니다.

yumkoori replied:
응답 코드를 따로 생성해서 Controller에서 사용하도록 수정했습니다.

yumkoori committed:
Refactor : 회원가입 함수중 예외 로직처리에서 가독성이 떨어지는 부분을 리팩토링함

- 기존 join함수에서 성공 로직이 catch 부분에서 처리되고, 실패로직은 try에서 처리되는 가독성 문제가 있었음
- 회원가입 여부 체크 로직을 따로 함수로 분리시켜서 기존 join함수 로직을 간결하게 리팩토링함

yumkoori commented on src/main/java/toyProject/toyProject01/member/application/service/MemberService.java:
```diff
@@ -64,4 +60,21 @@ public boolean Login(LoginCommand loginCommand) {
+        try {
+            Member findMember = loadMemberPort.loadMemberWithEmail(email);
+
+            if(findMember != null) {
+                log.info("존재하는 이메일 입니다.");
+            }
```
if문안에서 log로만 작성해둔 이유는 return false를 바로하면 깔끔해서 좋지만, return 문법 오류가 발생하여 밖에도 한번더 false를 반환해야 되는 문제가 있어서 일단 가독성 때문에 log라도 남겨두었습니다.

yumkoori commented on src/main/java/toyProject/toyProject01/member/application/service/MemberService.java:
```diff
@@ -29,19 +29,15 @@ public class MemberService implements MemberJoinUseCase, MemberLoginUseCase {
+
             return true;
+
+        } else {
+            throw new MemberServiceException("이미 가입된 회원입니다.");
```
회원가입 실패 테스트를 예외로 처리하고 있어서 회원가입이 불가능할때 false 대신에 예외를 반환시켰습니다.

yumkoori committed:
Feat : API 응답에 사용할 비즈니스 응답 코드를 새로 정의함

- 코드는 1000번대 코드 사용 예정
- 10xx번대 코드는 공통적인 응답코드 정의 예정
- 11xx번대 코드는 회원 서비스 관련 응답코드 정의
- 12xx번대 코드는 게시판 관련 응답코드 정의 예정

yumkoori committed:
Refactor : API Controller에서  비즈니스적인 에러 발생시 정의된 응답코드를 반환 하도록 수정함

- 회원가입 로직 성공시에는 단순히 Http스펙에 정의된 성공 코드를 반환함
- 회원가입 로직 실행중 비즈니스적인 예외 발생시 memberExHandle에서 예외를 처리하도록 분리시킴

yumkoori committed:
Refactor : LoginUsecase의 반환 타입을 Member로 변경

- Controller에서는 Login 성공시 응답데이터에 회원 닉네임 데이터를 담아 줘야하는데,
  클라이언트에서 넘어온 email, pw 정보밖에 없고, Service에서는 회원 정보를 반환해주지 않기 때문에 응답 데이터를 보낼 수 없다

- 또한 로직을 수행할때, boolean값을 이용할 용도가 없다

yumkoori committed:
Refator : ErrorCode에서 code필드 구조 재정의

- "M001"과 같은 code 데이터는 현재 의미가 없고, 필요 용도가 아니기때문에 삭제함

- 기존 code와 status가 의미적으로 헷갈릴 수 있기 때문에 Integer 타입을 code 필드로 재정의함

yumkoori committed:
Refactor : Service계층에서 예외를 던질때 오류코드를 포함하도록 수정함

- Controller에서 어떤 종류의 예외인지 구분할 수 있어야 자세한 예외 처리가 가능하기 때문에, Service에서 에러 코드를 추가로 보내주도록 수정함
- Service에서 사용하는 에러 코드에는 반드시 특정한 에러 코드가 포함되어야 하도록 수정함
- Servcie에서는 각 비즈니스 예외 상황에 따라서 에러코드를 선택해서 Controller로 알리도록 수정함

yumkoori committed:
Refactor : APIController를 Join 기능과 Login 기능으로 분리시키고 세부적인 예외처리를 추가함

- Join과 Login 기능의 Controller는 기능 로직과 의도가 다르고, 처리하는 예외 케이스도 다르기 때문에 분리함
- 비즈니스적인 예외 발생시 Service에서 넘어온 예외코드에 따라서 상응하는 응답 메세지를 반환하도록 수정함

yumkoori commented:
[ef2da16](https://github.com/yumkoori/toyProject01/pull/2/commits/ef2da1659922ef486e3a0d7b12e4fc6929bfa484) 이 부분 부터 새로 수정된 코드입니다.
